# Linear Regression Example
Here demonstrates the three core fitting methods of the `LinearRegression` model: OLS, Ridge, and Gradient Descent (GD), using a single synthetic dataset.

In [2]:
import numpy as np
from sklearn.datasets import make_regression 
from rice_ml.supervised_learning.linear_regression import LinearRegression
from rice_ml.processing.post_processing import r2_score
from rice_ml.processing.preprocessing import train_test_split


## 1. Load Data and Preparation
Generate a synthetic dataset suitable for regression, adding noise to make the problem realistic.

In [4]:
# Generate synthetic regression data: 100 samples, 5 features
X_raw, y = make_regression(
    n_samples = 100, 
    n_features = 5, 
    n_informative = 3, # 3 features are useful for prediction
    noise = 15.0, 
    random_state = 67
)

print(f"Total Samples: {X_raw.shape[0]}")
print(f"Number of Features: {X_raw.shape[1]}")

Total Samples: 100
Number of Features: 5


### 2. Data Pre-Processing: Scaling and Splitting
While OLS and Ridge (closed-form solutions) do not strictly require scaling, Gradient Descent requires it for convergence. We standardize the features to ensure all methods can be compared fairly.

In [6]:
# Standardization (Z-score scaling)
X_mean = np.mean(X_raw, axis = 0)
X_std = np.std(X_raw, axis = 0)
X_scaled = (X_raw - X_mean) / X_std

# Split the scaled dataset
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size = 0.3, random_state = 67
)

# Verification
print(f"\nTraining Set Size: {X_train.shape[0]} samples")


Training Set Size: 70 samples


### 3. Model Training and Evaluation
We train the model using all three modes: OLS, Ridge, and Gradient Descent.
#### 3.1 Ordinary Least Squares (OLS- Closed Form)
This model solves the Normal Equaton directly, giving the mathematically exact solution for unregularized linear regression.

In [7]:
# Initialize OLS model (method='ols')
ols_model = LinearRegression(method='ols')

print("Beginning OLS Training (Closed-Form Solution)...")
ols_model.fit(X_train, y_train)

# Evaluation
r2_ols = ols_model.score(X_test, y_test)

print("Training Complete.")
print(f"OLS R^2 Score: {r2_ols:.4f}")
print(f"Intercept (b_): {ols_model.b_:.2f}")

Beginning OLS Training (Closed-Form Solution)...
Training Complete.
OLS R^2 Score: 0.9728
Intercept (b_): -8.10


#### 4.2 Ridge Regression (Closed-Form with L2 Penalty)
Ridge adds a regularization penalty (`alpha`) to stabilize the solution and reduce the magnitude of the coefficients.

In [8]:
# Initialize Ridge model (method='ridge', setting alpha)
ridge_model = LinearRegression(method='ridge', alpha=10.0)

print("\nBeginning Ridge Training (Closed-Form with L2 Penalty, alpha=10.0)...")
ridge_model.fit(X_train, y_train)

# Evaluation
r2_ridge = ridge_model.score(X_test, y_test)

print("Training Complete.")
print(f"Ridge R^2 Score: {r2_ridge:.4f}")
# Note: Ridge coefficients should be slightly smaller than OLS coefficients


Beginning Ridge Training (Closed-Form with L2 Penalty, alpha=10.0)...
Training Complete.
Ridge R^2 Score: 0.9576


#### 4.3 Gradient Descent (GD - Iterative Solution)
GD iteratively updates weights using the learning rate (`eta`) for a fixed number of iterations (`epochs`). This requires feature scaling!

In [9]:
# Initialize GD model (method='gd', setting optimization parameters)
gd_model = LinearRegression(
    method='gd', 
    eta=0.01, 
    epochs=1000, 
    random_state=67
)

print("\nBeginning GD Training (Iterative Solution, 1000 epochs)...")
gd_model.fit(X_train, y_train)

# Evaluation
r2_gd = gd_model.score(X_test, y_test)

print("Training Complete.")
print(f"GD R^2 Score: {r2_gd:.4f}")

# You can inspect the convergence path
# print(f"Final Cost: {gd_model.cost_history_[-1]:.4f}")


Beginning GD Training (Iterative Solution, 1000 epochs)...
Training Complete.
GD R^2 Score: 0.9728


### 4. Summary of Results
| Method | Optimization Type | $R^2$ Score |
| :--- | :--- | :--- |
| **OLS (Ordinary Least Squares)** | Closed-form| 0.9728 |
| **Ridge Regression** | Closed-form (Regularized) |0.9576 |
| **GD (Gradient Descent)** | Iterative | 0.9728 |